<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Ivory Blakley">
  <link rel="shortcut icon" href="../../../img/favicon.ico">
  <title>Module.report.taxa - BioLockJ User Guide</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "Module.report.taxa";
    var mkdocs_page_input_path = "module/report/module.report.taxa.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../../.." class="icon icon-home"> BioLockJ User Guide</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../../Getting-Started/">Getting Started</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../../Commands/">Commands</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Pipeline Componenets</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../../Configuration/">Configuration</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../Built-in-modules/">Modules</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../The-Metadata-File/">Metadata</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../Dependencies/">Dependencies</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Features</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../../Check-Dependencies/">Check Dependencies</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../Failure-Recovery/">Failure Recovery</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../Validation/">Validation</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../Building-Modules/">Building Modules</a>
                </li>
                <li class="">
                    
    <a class="" href="../../../BioLockJ-Api/">BioLockJ API</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../../Example-Pipeline/">Example Pipeline</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../../FAQ/">FAQ</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../..">BioLockJ User Guide</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../..">Docs</a> &raquo;</li>
    
      
    
    <li>Module.report.taxa</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p>Modules in the <strong><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/report/package-summary.html">biolockj.module.report.taxa</a></strong> package process <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModule.html">ParserModule</a> output to produce or process taxa tables.</p>
<hr />
<h4 id="addmetadatatotaxatables">AddMetadataToTaxaTables</h4>
<p><code>#BioModule biolockj.module.report.taxa.AddMetadataToTaxaTables</code></p>
<p><strong>Description:</strong>  Map metadata onto taxa tables using sample ID.</p>
<p><strong>Options:</strong></p>
<ul>
<li><em>metadata.columnDelim</em> </li>
<li><em>metadata.commentChar</em></li>
<li><em>metadata.filePath</em></li>
<li><em>metadata.nullValue</em></li>
<li><em>report.taxonomyLevels</em></li>
</ul>
<hr />
<h4 id="addpseudocount">AddPseudoCount</h4>
<p><code>#BioModule biolockj.module.report.taxa.AddPseudoCount</code></p>
<p><strong>Description:</strong>  Add 1.0 to every value in each table.  This is typically done to avoid the mathmatecal consequences of 0's, and is generally only done on raw counts data.</p>
<p><strong>Options:</strong></p>
<p><em>none</em></p>
<hr />
<h4 id="buildtaxatables">BuildTaxaTables</h4>
<p><code>#BioModule biolockj.module.report.taxa.BuildTaxaTables</code></p>
<p><strong>Description:</strong>  Process <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModule.html">ParserModule</a> output to produce taxa tables.  This module reads the most recent OTU count files generated by any previous BioModule and re-writes the data as separate tables containing taxa counts for each taxonomy level.</p>
<p><strong>Options:</strong></p>
<ul>
<li><em>report.taxonomyLevels</em></li>
</ul>
<hr />
<h4 id="logtransformtaxatables">LogTransformTaxaTables</h4>
<p><code>#BioModule biolockj.module.report.taxa.LogTransformTaxaTables</code></p>
<p><strong>Description:</strong>  Log-transform the raw taxa counts on Log10 or Log-e scales.</p>
<p><strong>Options:</strong></p>
<ul>
<li><em>report.logBase</em> </li>
</ul>
<hr />
<h4 id="normalizebyreadspermillion">NormalizeByReadsPerMillion</h4>
<p><code>#BioModule biolockj.module.report.taxa.NormalizeByReadsPerMillion</code></p>
<p><strong>Description:</strong>  Normalize each sample for sequencing depth by reporting each value as the number of counts per million counts in a given sample.</p>
<p><strong>Options:</strong></p>
<p><em>none</em></p>
<hr />
<h4 id="normalizetaxatables">NormalizeTaxaTables</h4>
<p><code>#BioModule biolockj.module.report.taxa.NormalizeTaxaTables</code></p>
<p><strong>Description:</strong>  Normalize taxa tables based on formula:</p>
<p>
<script type="math/tex; mode=display"> counts_{normalized} = \frac{counts_{raw}}{n} \frac{\sum (x)}{N} +1 </script>
</p>
<p>Where:                                      </p>
<ul>
<li>
<script type="math/tex"> counts_{raw} </script> = raw count; the cell value before normalizing </li>
<li>
<script type="math/tex"> n </script> = number of sequences in the sample (total within a sample)</li>
<li>
<script type="math/tex"> \sum (x) </script> = total number of counts in the table (total across samples)</li>
<li>
<script type="math/tex"> N </script> = total number of samples</li>
</ul>
<p>Typically the data is put on a <script type="math/tex"> Log_{10} </script> scale, so the full forumula is:</p>
<p>
<script type="math/tex; mode=display"> counts_{final} = Log_{10} \biggl( \frac{counts_{raw}}{n} \frac{\sum (x)}{N} +1 \biggr) </script>
</p>
<p>The <script type="math/tex"> counts_{final} </script> values will be in output dir of the <code>LogTransformTaxaTables</code> module.  The <script type="math/tex"> counts_{normalized} </script> values will be in the output of the <code>NormalizeTaxaTables</code> module.</p>
<p>For further explanation regarding the normalization scheme, please read The ISME Journal 2013 paper by Dr. Anthony Fodor: <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3806260/">"Stochastic changes over time and not founder effects drive cage effects in microbial community assembly in a mouse model"</a></p>
<p>If report.logBase is not null, then the LogTransformTaxaTables will be added as a post-requisit module.</p>
<p><strong>Options:</strong></p>
<ul>
<li><em>report.logBase</em> </li>
</ul>
<hr />
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>2019 Fodor Lab</p>
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '../../..';</script>
    <script src="../../../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML" defer></script>
      <script src="../../../search/main.js" defer></script>

</body>
</html>
