Modules in the **[biolockj.module.report.taxa](https://msioda.github.io/BioLockJ/docs/biolockj/module/report/package-summary.html)** package process [ParserModule](https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModule.html) output to produce or process taxa tables.

---

#### AddMetadataToTaxaTables
`#BioModule biolockj.module.report.taxa.AddMetadataToTaxaTables`

**Description:**  Map metadata onto taxa tables using sample ID.

**Options:**

   - *metadata.columnDelim* 
   - *metadata.commentChar*
   - *metadata.filePath*
   - *metadata.nullValue*
   - *report.taxonomyLevels*

---

#### AddPseudoCount
`#BioModule biolockj.module.report.taxa.AddPseudoCount`

**Description:**  Add 1.0 to every value in each table.  This is typically done to avoid the mathmatecal consequences of 0's, and is generally only done on raw counts data.

**Options:**

*none*

---

#### BuildTaxaTables
`#BioModule biolockj.module.report.taxa.BuildTaxaTables`

**Description:**  Process [ParserModule](https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModule.html) output to produce taxa tables.  This module reads the most recent OTU count files generated by any previous BioModule and re-writes the data as separate tables containing taxa counts for each taxonomy level.

**Options:**

  - *report.taxonomyLevels*

---

#### LogTransformTaxaTables
`#BioModule biolockj.module.report.taxa.LogTransformTaxaTables`

**Description:**  Log-transform the raw taxa counts on Log10 or Log-e scales.

**Options:**

   - *report.logBase* 

---

#### NormalizeByReadsPerMillion
`#BioModule biolockj.module.report.taxa.NormalizeByReadsPerMillion`

**Description:**  Normalize each sample for sequencing depth by reporting each value as the number of counts per million counts in a given sample.

**Options:**

*none*

---

#### NormalizeTaxaTables
`#BioModule biolockj.module.report.taxa.NormalizeTaxaTables`

**Description:**  Normalize taxa tables based on formula:

$$ counts_{normalized} = \frac{counts_{raw}}{n} \frac{\sum (x)}{N} +1 $$

Where:                             			

* \( counts_{raw} \) = raw count; the cell value before normalizing 
* \( n \) = number of sequences in the sample (total within a sample)
* \( \sum (x) \) = total number of counts in the table (total across samples)
* \( N \) = total number of samples



Typically the data is put on a \( Log_{10} \) scale, so the full forumula is:

$$ counts_{final} = Log_{10} \biggl( \frac{counts_{raw}}{n} \frac{\sum (x)}{N} +1 \biggr) $$

The \( counts_{final} \) values will be in output dir of the `LogTransformTaxaTables` module.  The \( counts_{normalized} \) values will be in the output of the `NormalizeTaxaTables` module.


For further explanation regarding the normalization scheme, please read The ISME Journal 2013 paper by Dr. Anthony Fodor: ["Stochastic changes over time and not founder effects drive cage effects in microbial community assembly in a mouse model"](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3806260/)

If report.logBase is not null, then the LogTransformTaxaTables will be added as a post-requisit module.

**Options:**

   - *report.logBase* 

---
