BioLockJ requires [Java 1.8+](https://www.java.com "Java Homepage") and a [Unix-like](https://en.wikipedia.org/wiki/Unix-like) operating system  such as [Darwin/macOS](https://en.wikipedia.org/wiki/Darwin_(operating_system)), see [Notes about environments](../Supported-Environments).

BioLockJ is a pipeline manager, designed to integrate and manage external tools.  These external tools are not packaged into the BioLockJ program.  BioLockJ must run in an environment where these other tools have been installed **OR** run through docker using docker images that have the tools installed.  The core program, and all modules packaged with it, have corresponding docker images.

* Dependencies are required by modules listed in the **BioModule Function** column.
* Users **DO NOT NEED TO INSTALL** dependencies if not interested in the listed modules.
  - For example, if you intend to classify 16S samples with RDP and WGS samples with Kraken, do not install:<br>
    Bowtie2, GNU Awk, GNU Gzip, MetaPhlAn2, Python, QIIME 1, or Vsearch.

| # | Program | Version | BioModule Function | Link |
| :---: | :---: | :---: | :--- | :---: |
| 1 | Bowtie2 | 2.3.2 | [Metaphlan2Classifier](../module/classifier/module.classifier.wgs#metaphlan2classifier): Build reference indexes  | [download](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#obtaining-bowtie-2 "Bowtie2 Releases") |
| 2 | GNU Awk | 4.0.2 | [AwkFastaConverter](GENERATED/biolockj.module.seq/AwkFastaConverter.md): Convert Fastq to Fasta <br> [BuildQiimeMapping](../module/implicit/module.implicit.qiime#buildqiimemapping): Format metadata as QIIME mapping<br> [QiimeClosedRefClassifier](../module/classifier/module.classifier.r16s#qiimeclosedrefclassifier): Build batch mapping files | [download](https://www.gnu.org/software/gawk "Gawk Homepage")  |
| 3 | GNU Gzip | 1.5 | [AwkFastaConverter](GENERATED/biolockj.module.seq/AwkFastaConverter.md): Decompress .gz files <br> [Gunzipper](GENERATED/biolockj.module.seq/Gunzipper.md): Decompress .gz files | [download](https://www.gnu.org/software/gzip/ "Gzip Homepage") |
| 4 | Kraken | 0.10.5-beta | [KrakenClassifier](../module/classifier/module.classifier.wgs#krakenclassifier): Report WGS taxonomic summary | [download](https://ccb.jhu.edu/software/kraken "Kraken Homepage") |
| 5 | MetaPhlAn2 | 2.0 | [Metaphlan2Classifier](../module/classifier/module.classifier.wgs#metaphlan2classifier): Report WGS taxonomic summary (WGS) | [download](http://huttenhower.sph.harvard.edu/metaphlan2 "MetaPhlAn2 Homepage") |
| 6 | Python | 2.7.12 | [BuildQiimeMapping](../module/implicit/module.implicit.qiime#buildqiimemapping): Run validate_mapping_file.py<br> [MergeQiimeOtuTables](../module/implicit/module.implicit.qiime#MergeQiimeOtuTables): Run merge_otu_tables.py<br> [QiimeClosedRefClassifier](../module/classifier/module.classifier.r16s#qiimeclosedrefclassifier): Run pick_closed_reference_otus.py<br> [QiimeDeNovoClassifier](../module/classifier/module.classifier.r16s#qiimedenovoclassifier): Run pick_de_novo_otus.py<br> [QiimeOpenRefClassifier](../module/classifier/module.classifier.r16s#qiimeopenrefclassifier): Run pick_open_reference_otus.py<br> [QiimeClassifier](../module/implicit/module.implicit.qiime#QiimeClassifier): Run add_alpha_to_mapping_file.py, add_qiime_labels.py, alpha_diversity.py, filter_otus_from_otu_table.py, print_qiime_config.py, and summarize_taxa.py<br> [Metaphlan2Classifier](../module/classifier/module.classifier.wgs#metaphlan2classifier): Run metaphlan2.py | [download](https://www.python.org "Python Homepage") |
| 7 | PEAR | 0.9.8 | **Paired-End reAd merger**<br> [PearMergeReads](GENERATED/biolockj.module.seq/PearMergeReads.md) Merge paired Fastq files since some classifiers ([RDP](http://rdp.cme.msu.edu/classifier/classifier.jsp) & [QIIME](http://qiime.org)) will not accept paired reads. | [download](https://sco.h-its.org/exelixis/web/software/pear "PEAR Releases") |
| 8 | QIIME 1 | 1.9.1 |  **Quantitative Insights Into Microbial Ecology**<br> [BuildQiimeMapping](../module/implicit/module.implicit.qiime#buildqiimemapping): Validate QIIME mapping<br> [MergeQiimeOtuTables](../module/implicit/module.implicit.qiime#MergeQiimeOtuTables): Merge otu_table.biom files<br> [QiimeClosedRefClassifier](../module/classifier/module.classifier.r16s#qiimeclosedrefclassifier): Pick OTUs by reference<br> [QiimeDeNovoClassifier](../module/classifier/module.classifier.r16s#qiimedenovoclassifier): Pick OTUs by clustering <br> [QiimeOpenRefClassifier](../module/classifier/module.classifier.r16s#qiimeopenrefclassifier): Pick OTUs by reference and clustering<br> [QiimeClassifier](../module/implicit/module.implicit.qiime#QiimeClassifier): Report 16S taxonomic summary | [download](http://qiime.org "QIIME Homepage") |
| 9 | R | 3.5.0 | [R_CalculateStats](../module/report/module.report.r#r_calculatestats): Statistical modeling<br> [R_PlotPvalHistograms](../module/report/module.report.r#r_plotpvalhistograms): Plot p-value histograms for each reportable metadata field<br> [R_PlotOtus](../module/report/module.report.r#r_plototus): Build OTU-metadata boxplots and scatterplots<br> [R_PlotMds](../module/report/module.report.r#r_plotmds): Plot by top MDS axis <br> [R_PlotEffectSize](../module/report/module.report.r#r_ploteffectSize): Build barplot of effect magnetude by OTU/taxa | [download](https://cran.r-project.org "Comprehensive R Archive Network") |
| 10 | R-coin | 1.2 | **COnditional Inference procedures in a permutatioN test framework**<br> [R_CalculateStats](../module/report/module.report.r#r_calculatestats): Compute exact Wilcox_test  p-values | [download](https://cran.r-project.org/web/packages/coin/index.html "CRAN coin Pacakge Page") |
| 11 | R-ggpubr | 0.1.8 | [R_PlotPvalHistograms](../module/report/module.report.r#r_plotpvalhistograms): Set color palette <br> [R_PlotMds](../module/report/module.report.r#r_plotmds): Set color palette <br> [R_PlotEffectSize](../module/report/module.report.r#r_ploteffectSize): Set color palette | [download](https://cran.r-project.org/web/packages/ggpubr/index.html "CRAN ggpubr 'ggplot2' Based Publication Ready Plots") |
| 12 | R-Kendall | 2.2 | [R_CalculateStats](../module/report/module.report.r#r_calculatestats): Compute rank correlation p-values for continuous data types | [download](https://cran.r-project.org/web/packages/Kendall/index.html "CRAN Kendall Package Page") |
| 13 | R-properties | 0.0-9 | [R_Module](../module/report/module.report.r#r_module): Reads in the MASTER configuration properties file from the pipeline root directory | [download](https://cran.r-project.org/web/packages/properties/index.html "CRAN properties Package Page") |
| 14 | R-stringr | 1.2.0 | [R_Module](../module/report/module.report.r#r_module): For string manipulation for handling Configuration properties | [download](https://stringr.tidyverse.org/#installation "stringr Pacakge Page") |
| 15 | R-vegan | 2.5-2 | [R_PlotMds](../module/report/module.report.r#r_plotmds):  Ordination methods, diversity analysis and other functions for ecologists. | [download](https://cran.r-project.org/web/packages/vegan/index.html "CRAN vegan Package Page") |
| 16 | RDP | 2.12 | **Ribosomal Database Project**<br> [RdpClassifier](../module/classifier/module.classifier.r16s#RdpClassifier): Report 16S taxonomic summary  | [download](https://sourceforge.net/projects/rdp-classifier/files/rdp-classifier/rdp_classifier_2.12.zip/download "RDP Source Forge") |
| 17 | Vsearch | 2.4.3 | [QiimeDeNovoClassifier](../module/classifier/module.classifier.r16s#qiimedenovoclassifier): Chimera detection<br> [QiimeOpenRefClassifier](../module/classifier/module.classifier.r16s#qiimeopenrefclassifier): Chimera detection | [download](https://github.com/torognes/vsearch#download-and-install "Vsearch GitHub") |

### Version Dependencies
* The **Version** column contains the version tested during BioLockJ development, but other versions can often be substituted.  
* Major releases (such as Python 2 vs. Python 3) contain API changes that will not integrate with the current BioLockJ code.
* Application APIs often change over time, so not all versions are supported.  For example, Bowtie2 did not add the large index functionality until version 2.3.2. 