<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Ivory Blakley">
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>All Modules - BioLockJ User Guide</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "All Modules";
    var mkdocs_page_input_path = "GENERATED/all-modules.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> BioLockJ User Guide</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../Getting-Started/">Getting Started</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Pipeline Componenets</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../Configuration/">Configuration</a>
                </li>
                <li class="">
                    
    <a class="" href="../General-Properties/">General Properties</a>
                </li>
                <li class="">
                    
    <a class="" href="../../Built-in-modules/">Modules</a>
                </li>
                <li class="">
                    
    <a class="" href="../../The-Metadata-File/">Metadata</a>
                </li>
                <li class="">
                    
    <a class="" href="../../Dependencies/">Dependencies</a>
                </li>
                <li class="">
                    
    <span class="caption-text">Configuration Details</span>
    <ul class="subnav">
                <li class="toctree-l3">
                    
    <a class="" href="../Cluster/">cluster</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../Docker/">docker</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../Input/">input</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../Metadata/">metadata</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../R/">r</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../Report.md">report</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../Script/">script</a>
                </li>
                <li class="toctree-l3">
                    
    <a class="" href="../Validation/">validation</a>
                </li>
    </ul>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Features</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../Commands/">Commands</a>
                </li>
                <li class="">
                    
    <a class="" href="../../Check-Dependencies/">Check Dependencies</a>
                </li>
                <li class="">
                    
    <a class="" href="../../Failure-Recovery/">Failure Recovery</a>
                </li>
                <li class="">
                    
    <a class="" href="../Validation/">Validation</a>
                </li>
                <li class="">
                    
    <a class="" href="../../Building-Modules/">Building Modules</a>
                </li>
                <li class="">
                    
    <a class="" href="../BioLockJ-Api/">BioLockJ API</a>
                </li>
                <li class="">
                    
    <a class="" href="../biolockj-help/">BioLockJ help menu</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../Supported-Environments/">Environments</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../Example-Pipeline/">Example Pipeline</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../FAQ/">FAQ</a>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">BioLockJ User Guide</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
    
    <li>All Modules</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="all-modules">All Modules<a class="headerlink" href="#all-modules" title="Permanent link">#</a></h1>
<p><em>This is an auto-generated list of all modules with links to auto-generated module documentation.</em></p>
<p><a href="../biolockj.module.report.humann2/AddMetadataToPathwayTables/">AddMetadataToPathwayTables</a>                 <br />
<a href="../biolockj.module.report.taxa/AddMetadataToTaxaTables/">AddMetadataToTaxaTables</a>                 <br />
<a href="../biolockj.module.report.taxa/AddPseudoCount/">AddPseudoCount</a> - <em>Add a pseudocount (+1) to each value in each taxa table.</em>                 <br />
<a href="../biolockj.module.seq/AwkFastaConverter/">AwkFastaConverter</a> - <em>Convert fastq files into fasta format.</em>                 <br />
<a href="../biolockj.module.implicit.qiime/BuildQiimeMapping/">BuildQiimeMapping</a>                 <br />
<a href="../biolockj.module.report.taxa/BuildTaxaTables/">BuildTaxaTables</a>                 <br />
<a href="../biolockj.module.report.otu/CompileOtuCounts/">CompileOtuCounts</a>                 <br />
<a href="../biolockj.module.rnaseq/DESeq2/">DESeq2</a> - <em>Determine statistically significant differences using DESeq2.</em>                 <br />
<a href="../biolockj.module.implicit/Demultiplexer/">Demultiplexer</a> - <em>Demultiplex samples into separate files for each sample.</em>                 <br />
<a href="../biolockj.module.report/Email/">Email</a> - <em>Send an email containing the pipeline summary when the pipeline either completes or fails.</em>                 <br />
<a href="../biolockj.module.diy/GenMod/">GenMod</a> - <em>Allows user to add their own scripts into the BioLockJ pipeline.</em>                 <br />
<a href="../biolockj.module.assembly/GenomeAssembly/">GenomeAssembly</a> - <em>Assemble WGS sequences with MetaSPAdes, bin contigs with Metabat2 and check quality with checkM.</em>                 <br />
<a href="../biolockj.module.seq/Gunzipper/">Gunzipper</a> - <em>Decompress gzipped files.</em>                 <br />
<a href="../biolockj.module.classifier.wgs/Humann2Classifier/">HUMAnN2</a> - <em>Profile the presence/absence and abundance of microbial pathways in a community from metagenomic or metatranscriptomic sequencing data.</em>                 <br />
<a href="../biolockj.module.implicit.parser.wgs/Humann2Parser/">Humann2Parser</a> - <em>Build OTU tables from HumanN2 classifier module output.</em>                 <br />
<a href="../biolockj.module.implicit/ImportMetadata/">ImportMetadata</a> - <em>Read existing metadata file, or create a default one.</em>                 <br />
<a href="../biolockj.module.report/JsonReport/">JsonReport</a>                 <br />
<a href="../biolockj.module.seq/KneadData/">KneadData</a> - <em>Run the Biobakery <a href="https://bitbucket.org/biobakery/kneaddata/wiki/Home">KneadData</a> program to remove contaminated DNA.</em>                 <br />
<a href="../biolockj.module.classifier.wgs/Kraken2Classifier/">Kraken2Classifier</a> - <em>Classify WGS samples with <a href="https://ccb.jhu.edu/software/kraken2/">KRAKEN 2</a>.</em>                 <br />
<a href="../biolockj.module.implicit.parser.wgs/Kraken2Parser/">Kraken2Parser</a> - <em>Build OTU tables from <a href="http://ccb.jhu.edu/software/kraken/">KRAKEN</a> mpa-format reports.</em>                 <br />
<a href="../biolockj.module.classifier.wgs/KrakenClassifier/">KrakenClassifier</a> - <em>Classify WGS samples with KRAKEN.</em>                 <br />
<a href="../biolockj.module.implicit.parser.wgs/KrakenParser/">KrakenParser</a> - <em>Build OTU tables from <a href="http://ccb.jhu.edu/software/kraken/">KRAKEN</a> mpa-format reports.</em>                 <br />
<a href="../biolockj.module.report.taxa/LogTransformTaxaTables/">LogTransformTaxaTables</a>                 <br />
<a href="../biolockj.module.implicit.qiime/MergeQiimeOtuTables/">MergeQiimeOtuTables</a>                 <br />
<a href="../biolockj.module.classifier.wgs/Metaphlan2Classifier/">Metaphlan2Classifier</a> - <em>Classify WGS samples with <a href="http://bitbucket.org/biobakery/metaphlan2">MetaPhlAn2</a>.</em>                 <br />
<a href="../biolockj.module.implicit.parser.wgs/Metaphlan2Parser/">Metaphlan2Parser</a>                 <br />
<a href="../biolockj.module.seq/Multiplexer/">Multiplexer</a> - <em>Multiplex samples into a single file, or two files (one with forward reads, one with reverse reads) if multiplexing paired reads.</em>                 <br />
<a href="../biolockj.module.report.taxa/NormalizeByReadsPerMillion/">NormalizeByReadsPerMillion</a> - <em>Represent each count value in the table as: counts / (total counts in sample / 1 million)</em>                 <br />
<a href="../biolockj.module.report.taxa/NormalizeTaxaTables/">NormalizeTaxaTables</a>                 <br />
<a href="../biolockj.module.seq/PearMergeReads/">PearMergeReads</a> - <em>Run pear, the Paired-End reAd mergeR</em>                 <br />
<a href="../biolockj.module.implicit.qiime/QiimeClassifier/">QiimeClassifier</a>                 <br />
<a href="../biolockj.module.classifier.r16s/QiimeClosedRefClassifier/">QiimeClosedRefClassifier</a> - <em>Pick OTUs using a closed reference database and construct an OTU table via the QIIME script pick_closed_reference_otus.py</em>                 <br />
<a href="../biolockj.module.classifier.r16s/QiimeDeNovoClassifier/">QiimeDeNovoClassifier</a> - <em>Run the QIIME pick_de_novo_otus.py script on all fasta sequence files</em>                 <br />
<a href="../biolockj.module.classifier.r16s/QiimeOpenRefClassifier/">QiimeOpenRefClassifier</a> - <em>Run the QIIME pick_open_reference_otus.py script on all fasta sequence files</em>                 <br />
<a href="../biolockj.module.implicit.parser.r16s/QiimeParser/">QiimeParser</a>                 <br />
<a href="../biolockj.module.report.r/R_CalculateStats/">R_CalculateStats</a> - <em>Generate a summary statistics table with [adjusted and unadjusted] [parameteric and non-parametirc] p-values and r<sup>2</sup> values for each reportable metadata field and each </em>report.taxonomyLevel<em> configured.</em>                 <br />
<a href="../biolockj.module.report.r/R_PlotEffectSize/">R_PlotEffectSize</a> - <em>Generate horizontal barplot representing effect size (Cohen's d, r<sup>2</sup>, and/or fold change) for each reportable metadata field and each </em>report.taxonomyLevel<em> configured.</em>                 <br />
<a href="../biolockj.module.report.r/R_PlotMds/">R_PlotMds</a> - <em>Generate sets of multidimensional scaling plots showing 2 axes at a time (up to the &lt;</em>r_PlotMds.numAxis<em>&gt;th axis) with color coding based on each categorical metadata field (default) or by each field given in </em>r_PlotMds.reportFields<em><em>                 <br />
<a href="../biolockj.module.report.r/R_PlotOtus/">R_PlotOtus</a> - </em>Generate OTU-metadata box-plots and scatter-plots for each reportable metadata field and each </em>report.taxonomyLevel<em> configured</em>                 <br />
<a href="../biolockj.module.report.r/R_PlotPvalHistograms/">R_PlotPvalHistograms</a> - <em>Generate p-value histograms for each reportable metadata field and each </em>report.taxonomyLevel<em> configured</em>                 <br />
<a href="../biolockj.module.report.otu/RarefyOtuCounts/">RarefyOtuCounts</a> - <em>Applies a mean iterative post-OTU classification rarefication algorithm so that each output sample will have approximately the same number of OTUs.</em>                 <br />
<a href="../biolockj.module.seq/RarefySeqs/">RarefySeqs</a> - <em>Randomly sub-sample sequences to reduce all samples to the configured maximum.</em>                 <br />
<a href="../biolockj.module.classifier.r16s/RdpClassifier/">RdpClassifier</a> - <em>Classify 16s samples with <a href="http://rdp.cme.msu.edu/classifier/classifier.jsp">RDP</a>.</em>                 <br />
<a href="../biolockj.module.implicit.parser.r16s/RdpParser/">RdpParser</a> - <em>Build OTU tables from <a href="http://rdp.cme.msu.edu/classifier/classifier.jsp">RDP</a> reports.</em>                 <br />
<a href="../biolockj.module.implicit/RegisterNumReads/">RegisterNumReads</a>                 <br />
<a href="../biolockj.module.report.otu/RemoveLowOtuCounts/">RemoveLowOtuCounts</a>                 <br />
<a href="../biolockj.module.report.humann2/RemoveLowPathwayCounts/">RemoveLowPathwayCounts</a>                 <br />
<a href="../biolockj.module.report.otu/RemoveScarceOtuCounts/">RemoveScarceOtuCounts</a>                 <br />
<a href="../biolockj.module.report.humann2/RemoveScarcePathwayCounts/">RemoveScarcePathwayCounts</a>                 <br />
<a href="../biolockj.module.seq/SeqFileValidator/">SeqFileValidator</a> - <em>This BioModule validates fasta/fastq file formats are valid and enforces min/max read lengths.</em>                 <br />
<a href="../biolockj.module.getData.sra/SraDownload/">SraDownload</a> - <em>SraDownload downloads and compresses short read archive (SRA) files to fastq.gz</em>                 <br />
<a href="../biolockj.module.getData.sra/SraMetaDB/">SraMetaDB</a> - <em>Makes sure that the SRAmetadb exists, downloads if it does not already exist.</em>                 <br />
<a href="../biolockj.module.getData.sra/SraMetaData/">SraMetaData</a> - <em>Extract metadata via pysradb from local copy of SRAmetadb.sqlite.</em>                 <br />
<a href="../biolockj.module.getData.sra/SrpSrrConverter/">SrpSrrConverter</a> - <em>Create an SraAccList.txt file from an SRA project identifier.</em>                 <br />
<a href="../biolockj.module.seq/TrimPrimers/">TrimPrimers</a> - <em>Remove primers from reads, option to discard reads unless primers are attached to both forward and reverse reads.</em>                   </p>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
      <p>2019 Fodor Lab</p>
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML" defer></script>
      <script src="../../search/main.js" defer></script>

</body>
</html>
