



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="https://ivoryc.github.io/BioLockJ/Building-Modules/">
      
      
        <meta name="author" content="Ivory Blakley">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.5.1">
    
    
      
        <title>Building Modules - BioLockJ User Guide</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.1b62728e.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.268332fc.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
    
      
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#any-java-class-that-implements-the-biomodule-interface-can-be-added-to-a-biolockj-pipeline" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://ivoryc.github.io/BioLockJ" title="BioLockJ User Guide" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            <span class="md-header-nav__topic">
              BioLockJ User Guide
            </span>
            <span class="md-header-nav__topic">
              
                Building Modules
              
            </span>
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  

<a href="https://github.com/BioLockJ-Dev-Team/BioLockJ/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    BioLockJ
  </div>
</a>
          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main" role="main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="https://ivoryc.github.io/BioLockJ" title="BioLockJ User Guide" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    BioLockJ User Guide
  </label>
  
    <div class="md-nav__source">
      


  

<a href="https://github.com/BioLockJ-Dev-Team/BioLockJ/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    BioLockJ
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Installation/" title="Installation" class="md-nav__link">
      Installation
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Getting-Started/" title="Getting Started" class="md-nav__link">
      Getting Started
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Commands/" title="Commands" class="md-nav__link">
      Commands
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-5" type="checkbox" id="nav-5">
    
    <label class="md-nav__link" for="nav-5">
      Pipeline Componenets
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-5">
        Pipeline Componenets
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../Configuration/" title="Configuration" class="md-nav__link">
      Configuration
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../Built-in-modules/" title="Modules" class="md-nav__link">
      Modules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../The-Metadata-File/" title="Metadata" class="md-nav__link">
      Metadata
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../Dependencies/" title="Dependencies" class="md-nav__link">
      Dependencies
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-6" type="checkbox" id="nav-6" checked>
    
    <label class="md-nav__link" for="nav-6">
      Features
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-6">
        Features
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../Failure-Recovery/" title="Failure Recovery" class="md-nav__link">
      Failure Recovery
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../Validation/" title="Validation" class="md-nav__link">
      Validation
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Building Modules
      </label>
    
    <a href="./" title="Building Modules" class="md-nav__link md-nav__link--active">
      Building Modules
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#any-java-class-that-implements-the-biomodule-interface-can-be-added-to-a-biolockj-pipeline" class="md-nav__link">
    Any Java class that implements the BioModule interface can be added to a BioLockJ pipeline.
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#beginners" class="md-nav__link">
    Beginners
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#coding-your-module" class="md-nav__link">
    Coding your module
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#to-support-a-new-classifier-create-3-modules-that-implement-the-following-interfaces" class="md-nav__link">
    To support a new classifier, create 3 modules that implement the following interfaces:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#biomoduleimpl-is-the-top-level-superclass-for-all-modules" class="md-nav__link">
    BioModuleImpl is the top-level superclass for all modules.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scriptmoduleimpl-extends-biomoduleimpl-superclass-for-script-generating-modules" class="md-nav__link">
    ScriptModuleImpl extends BioModuleImpl:  superclass for script-generating modules.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#javamoduleimpl-extends-scriptmoduleimpl-superclass-for-pure-java-modules" class="md-nav__link">
    JavaModuleImpl extends ScriptModuleImpl: superclass for pure Java modules.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#classifiermoduleimpl-extends-scriptmoduleimpl-biolockjmoduleclassifier-superclass" class="md-nav__link">
    ClassifierModuleImpl extends ScriptModuleImpl: biolockj.[[module.classifier]] superclass.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parsermoduleimpl-extends-javamoduleimpl-biolockjmoduleimplicitparser-superclass" class="md-nav__link">
    ParserModuleImpl extends JavaModuleImpl: biolockj.[[module.implicit.parser]] superclass.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#otunodeimpl-is-the-superclass-for-the-biolockjnode-package" class="md-nav__link">
    OtuNodeImpl is the superclass for the biolockj.node package.
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#share-your-module" class="md-nav__link">
    Share your module
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../Example-Pipeline/" title="Example Pipeline" class="md-nav__link">
      Example Pipeline
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../FAQ/" title="FAQ" class="md-nav__link">
      FAQ
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-9" type="checkbox" id="nav-9">
    
    <label class="md-nav__link" for="nav-9">
      more
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-9">
        more
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../module/classifier/module.classifier/" title="Classifier Package" class="md-nav__link">
      Classifier Package
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/classifier/module.classifier.r16s/" title="Module.classifier.r16s" class="md-nav__link">
      Module.classifier.r16s
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/classifier/module.classifier.wgs/" title="Whole Genome Sequence Classifiers" class="md-nav__link">
      Whole Genome Sequence Classifiers
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/diy/module.diy/" title="Do It Yourself - DIY Package" class="md-nav__link">
      Do It Yourself - DIY Package
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/implicit/module.implicit/" title="Module.implicit" class="md-nav__link">
      Module.implicit
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/implicit/module.implicit.parser/" title="Module.implicit.parser" class="md-nav__link">
      Module.implicit.parser
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/implicit/module.implicit.parser.r16s/" title="Module.implicit.parser.r16s" class="md-nav__link">
      Module.implicit.parser.r16s
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/implicit/module.implicit.parser.wgs/" title="Module.implicit.parser.wgs" class="md-nav__link">
      Module.implicit.parser.wgs
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/implicit/module.implicit.qiime/" title="Module.implicit.qiime" class="md-nav__link">
      Module.implicit.qiime
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/report/module.report.humann2/" title="Pathway Modules" class="md-nav__link">
      Pathway Modules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/report/module.report/" title="Report Package" class="md-nav__link">
      Report Package
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/report/module.report.otu/" title="OTU report modules" class="md-nav__link">
      OTU report modules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/report/module.report.r/" title="R Report Modules" class="md-nav__link">
      R Report Modules
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/report/module.report.taxa/" title="Module.report.taxa" class="md-nav__link">
      Module.report.taxa
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../module/seq/module.seq/" title="Seq Package" class="md-nav__link">
      Seq Package
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#any-java-class-that-implements-the-biomodule-interface-can-be-added-to-a-biolockj-pipeline" class="md-nav__link">
    Any Java class that implements the BioModule interface can be added to a BioLockJ pipeline.
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#beginners" class="md-nav__link">
    Beginners
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#coding-your-module" class="md-nav__link">
    Coding your module
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#to-support-a-new-classifier-create-3-modules-that-implement-the-following-interfaces" class="md-nav__link">
    To support a new classifier, create 3 modules that implement the following interfaces:
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#biomoduleimpl-is-the-top-level-superclass-for-all-modules" class="md-nav__link">
    BioModuleImpl is the top-level superclass for all modules.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scriptmoduleimpl-extends-biomoduleimpl-superclass-for-script-generating-modules" class="md-nav__link">
    ScriptModuleImpl extends BioModuleImpl:  superclass for script-generating modules.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#javamoduleimpl-extends-scriptmoduleimpl-superclass-for-pure-java-modules" class="md-nav__link">
    JavaModuleImpl extends ScriptModuleImpl: superclass for pure Java modules.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#classifiermoduleimpl-extends-scriptmoduleimpl-biolockjmoduleclassifier-superclass" class="md-nav__link">
    ClassifierModuleImpl extends ScriptModuleImpl: biolockj.[[module.classifier]] superclass.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parsermoduleimpl-extends-javamoduleimpl-biolockjmoduleimplicitparser-superclass" class="md-nav__link">
    ParserModuleImpl extends JavaModuleImpl: biolockj.[[module.implicit.parser]] superclass.
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#otunodeimpl-is-the-superclass-for-the-biolockjnode-package" class="md-nav__link">
    OtuNodeImpl is the superclass for the biolockj.node package.
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#share-your-module" class="md-nav__link">
    Share your module
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
  
    <h1>Building Modules</h1>
  
  <h4 id="any-java-class-that-implements-the-biomodule-interface-can-be-added-to-a-biolockj-pipeline">Any Java class that implements the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/BioModule.html">BioModule</a> interface can be added to a BioLockJ pipeline.</h4>
<p>The BioLockJ v1.0 implementation is currently focused on metagenomics analysis, but the generalized application framework is not limited to this domain.  Users can implement new BioModules to automate a wide variety of bioinformatics and report analytics.  The <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/BioModule.html">BioModule</a> interface was designed so that users can develop new modules on their own.  </p>
<h2 id="beginners">Beginners</h2>
<p>See the BioModule hello world tutorial.</p>
<h2 id="coding-your-module">Coding your module</h2>
<p>To create a new <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/BioModule.html">BioModule</a>, simply extend one of the abstract Java superclasses, code it's abstract methods, and add it to your pipeline with #BioModule tag your Config file:<br>
1. <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/BioModuleImpl.html">BioModuleImpl</a>: Extend if a more specific interface does not apply
1. <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/ScriptModuleImpl.html">ScriptModuleImpl</a>: Extend if your module generates and executes bash scripts
1. <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/JavaModuleImpl.html">JavaModuleImpl</a>: Extend if your module only runs Java code 
1. <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/classifier/ClassifierModuleImpl.html">ClassifierModuleImpl</a>: Extend to support a new classifier program
1. <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModuleImpl.html">ParserModuleImpl</a>: Extend to parse output of a new classifier program
1. <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/report/r/R_Module.html">R_Module</a>: Extend if your module generates and executes R scripts</p>
<h4 id="to-support-a-new-classifier-create-3-modules-that-implement-the-following-interfaces">To support a new classifier, create 3 modules that implement the following interfaces:</h4>
<ol>
<li><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/classifier/ClassifierModule.html">ClassifierModule</a>: Implement to generate bash scripts needed to call classifier program</li>
<li><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModule.html">ParserModule</a>: Implement to parse classifier output, configure as classifier post-requisite </li>
<li><a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/OtuNode.html">OtuNode</a>: Classifier specific implementation holds OTU information for 1 sequence</li>
</ol>
<h4 id="biomoduleimpl-is-the-top-level-superclass-for-all-modules"><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/BioModuleImpl.html">BioModuleImpl</a> is the top-level superclass for all modules.</h4>
<table>
<thead>
<tr>
<th align="left">Method</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><strong>checkDependencies()</strong></td>
<td align="left"><strong>Must override.</strong>  Called before executeTask() to identify [[Configuration]] errors and perform runtime validations.</td>
</tr>
<tr>
<td align="left"><strong>executeTask()</strong></td>
<td align="left"><strong>Must override.</strong>  Executes core module logic.</td>
</tr>
<tr>
<td align="left">cleanUp()</td>
<td align="left">Called after executeTask() to run cleanup operations, update Config properties, etc.</td>
</tr>
<tr>
<td align="left">getInputFiles()</td>
<td align="left">Return previous module output.</td>
</tr>
<tr>
<td align="left">getModuleDir()</td>
<td align="left">Return module root directory.</td>
</tr>
<tr>
<td align="left">getOutputDir()</td>
<td align="left">Return module output directory.</td>
</tr>
<tr>
<td align="left">getPostRequisiteModules()</td>
<td align="left">Returns a list of BioModules to run after the current module.</td>
</tr>
<tr>
<td align="left">getPreRequisiteModules()</td>
<td align="left">Returns a list of BioModules to run before the current module.</td>
</tr>
<tr>
<td align="left">getSummary()</td>
<td align="left">Return output directory summary.  Most modules override this method by adding module specific summary details to super.getSummary().</td>
</tr>
<tr>
<td align="left">getTempDir()</td>
<td align="left">Return module temp directory.</td>
</tr>
<tr>
<td align="left">setModuleDir(path)</td>
<td align="left">Set module directory.</td>
</tr>
</tbody>
</table>
<h4 id="scriptmoduleimpl-extends-biomoduleimpl-superclass-for-script-generating-modules"><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/ScriptModuleImpl.html">ScriptModuleImpl</a> extends <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/BioModuleImpl.html">BioModuleImpl</a>:  superclass for script-generating modules.</h4>
<table>
<thead>
<tr>
<th align="left">Method</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><strong>buildScript(files)</strong></td>
<td align="left"><strong>Must override.</strong>  Called by executeTask() for datasets with forward reads only.  The return type is a list of lists.  Each nested list contains the bash script lines required to process 1 sample.  Obtains sequence files from getInputFiles().</td>
</tr>
<tr>
<td align="left">buildScriptForPairedReads(files)</td>
<td align="left">Calls back to buildScript(files) by default.  Subclasses override this method to generate unique scripts for datasets containing paired reads.</td>
</tr>
<tr>
<td align="left">checkDependencies()</td>
<td align="left">Called before executeTask() to validate <em>script.batchSize</em>, <em>script.exitOnError</em>, <em>script.numThreads</em>, <em>script.permissions</em>, <em>script.timeout</em></td>
</tr>
<tr>
<td align="left">getJobParams()</td>
<td align="left">Return shell command to execute the MAIN script.</td>
</tr>
<tr>
<td align="left">getScriptDir()</td>
<td align="left">Return module script directory.</td>
</tr>
<tr>
<td align="left">getSummary()</td>
<td align="left">Adds the script directory summary to super.getSummary().  Most modules override this method by adding module specific summary details to super.getSummary().</td>
</tr>
<tr>
<td align="left">getTimeout()</td>
<td align="left">Return <em>script.timeout</em>.</td>
</tr>
<tr>
<td align="left">getWorkerScriptFunctions()</td>
<td align="left">Return bash script lines for any functions needed in the worker scripts.</td>
</tr>
</tbody>
</table>
<h4 id="javamoduleimpl-extends-scriptmoduleimpl-superclass-for-pure-java-modules"><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/JavaModuleImpl.html">JavaModuleImpl</a> extends <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/ScriptModuleImpl.html">ScriptModuleImpl</a>: superclass for pure Java modules.</h4>
<p>To avoid running code on the cluster head node, a temporary instance of BioLockJ is spawned on a cluster node which is launched by the sole worker script from the job queue. </p>
<table>
<thead>
<tr>
<th align="left">Method</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><strong>runModule()</strong></td>
<td align="left"><strong>Must override.</strong>  Executes core module logic.</td>
</tr>
<tr>
<td align="left">buildScript(files)</td>
<td align="left">This method returns a single line calling java on the BioLockJ source code, passing -d parameter to run in direct mode and the full class name of the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/JavaModule.html">JavaModule</a> to indicate the module to run.</td>
</tr>
<tr>
<td align="left">getSource()</td>
<td align="left">Determines if running code from Jar or source code in order to write valid bash script lines.</td>
</tr>
<tr>
<td align="left">getTimeout()</td>
<td align="left">Return <em>java.timeout</em>.</td>
</tr>
<tr>
<td align="left">moduleComplete()</td>
<td align="left">Create the script success indicator file.</td>
</tr>
<tr>
<td align="left">moduleFailed()</td>
<td align="left">Create the script failures indicator file.</td>
</tr>
</tbody>
</table>
<h4 id="classifiermoduleimpl-extends-scriptmoduleimpl-biolockjmoduleclassifier-superclass"><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/classifier/ClassifierModuleImpl.html">ClassifierModuleImpl</a> extends <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/ScriptModuleImpl.html">ScriptModuleImpl</a>: biolockj.[[module.classifier]] superclass.</h4>
<table>
<thead>
<tr>
<th align="left">Method</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">buildScriptForPairedReads(files)</td>
<td align="left">Called by executeTask() for datasets with paired reads.  The return type is a list of lists, where each nested list contains the bash script lines required to process 1 sample.  Obtains sequence files from <a href="https://msioda.github.io/BioLockJ/docs/biolockj/util/SeqUtil.html">SeqUtil</a>.getPairedReads(getInputFiles()).</td>
</tr>
<tr>
<td align="left">checkDependencies()</td>
<td align="left">Validate [[Configuration]] properties <em>exe.classifier</em> and <em>exe.classifierParams</em>, verify sequence file format, log classifier version info, and verify no biolockj.[[module.seq]] modules are configured run after the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/classifier/ClassifierModule.html">ClassifierModule</a>. Subclasses should call super.checkDependencies() if overriding this method to retain these verifications.</td>
</tr>
<tr>
<td align="left">executeTask()</td>
<td align="left">Call buildScript(files) or buildScriptForPairedReads(files) based input sequence format and calls <a href="https://msioda.github.io/BioLockJ/docs/biolockj/util/BashScriptBuilder.html">BashScriptBuilder</a> to generate the main script + 1 worker script for every <em>script.batchSize</em> samples.  To change the batch scheme, override this method to call the alternate <a href="https://msioda.github.io/BioLockJ/docs/biolockj/util/BashScriptBuilder.html">BashScriptBuilder</a>.buildScripts() method signiture and hard code the batch size.  All biolockj.[[module.classifier]] modules override this method.</td>
</tr>
<tr>
<td align="left">getClassifierExe()</td>
<td align="left">Return [[Configuration]] property <em>exe.classifier</em> to call the classifier program in the bash scripts.  If the classifier is not included in <em>cluster.modules</em>, validate that value is a valid file path.  If <em>exe.classifier</em> is undefined, replace the property prefix <em>exe</em> with the lowercase prefix of the module class name (less the standard module suffix <em>classifier</em>).  For example, use <em>rdp.classifier</em> for <a href="module.classifier.r16s#RdpClassifier">RdpClassifier</a> and <em>kraken.classifier</em> for <a href="module.classifier.wgs#KrakenClassifier">KrakenClassifier</a>.  This allows users to define all classifier programs in a default [[Configuration]] file rather than setting <em>exe.clssifier</em> in each project [[Configuration]] file.</td>
</tr>
<tr>
<td align="left">getClassifierParams()</td>
<td align="left">Return [[Configuration]] property <em>exe.classifierParams</em> which may contain a list of parameters (without hyphens) to pass to the classifier program in the bash scripts.  If <em>exe.classifierParams</em> is undefined, replace the property prefix <em>exe</em> with the lowercase prefix of the module class name as described for <em>exe.classifier</em>.</td>
</tr>
<tr>
<td align="left">getSummary()</td>
<td align="left">Adds input directory summary to super.getSummary().  Most modules override this method to add module specific summary details to super.getSummary().</td>
</tr>
<tr>
<td align="left">logVersion()</td>
<td align="left">Run <em>exe.classifier</em> --version to log version info.  <a href="module.classifier.r16s#RdpClassifier">RDP</a> overrides this method to return null since the version switch is not supported.</td>
</tr>
</tbody>
</table>
<h4 id="parsermoduleimpl-extends-javamoduleimpl-biolockjmoduleimplicitparser-superclass"><a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModuleImpl.html">ParserModuleImpl</a> extends <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/JavaModuleImpl.html">JavaModuleImpl</a>: biolockj.[[module.implicit.parser]] superclass.</h4>
<table>
<thead>
<tr>
<th align="left">Method</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left"><strong>parseSamples()</strong></td>
<td align="left"><strong>Must override.</strong>  Called by executeTask() to populate the Set returned by getParsedSamples().  Each classifier requires a unique parser module to decode its output.  This method should iterate through the classifier reports to build <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/OtuNode.html">OtuNode</a>s for each sample-OTU found in the report.  The <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/OtuNode.html">OtuNode</a>s are stored in a <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/ParsedSample.html">ParsedSample</a> and cached via addParsedSample(<a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/ParsedSample.html">ParsedSample</a>).</td>
</tr>
<tr>
<td align="left">addParsedSample( sample )</td>
<td align="left">Add the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/ParsedSample.html">ParsedSample</a> to the Set returned by getParsedSamples().</td>
</tr>
<tr>
<td align="left">buildOtuTables()</td>
<td align="left">Generate OTU abundance tables from <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/classifier/ClassifierModule.html">ClassifierModule</a> output.</td>
</tr>
<tr>
<td align="left">checkDependencies()</td>
<td align="left">Validate [[Configuration]] properties (<em>report.minOtuCount</em>, <em>report.minOtuThreshold</em>, <em>report.logBase</em>) and  verify no biolockj.[[module.classifier]] modules are configured to run after the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/module/implicit/parser/ParserModule.html">ParserModule</a>.</td>
</tr>
<tr>
<td align="left">executeTask()</td>
<td align="left">If <em>report.numHits</em>=Y, add "Num_Hits" column to metadata containing the number of reads that map to any OTU for each sample.  Calls buildOtuTables() to generate module output.</td>
</tr>
<tr>
<td align="left">getParsedSample(id)</td>
<td align="left">Return the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/ParsedSample.html">ParsedSample</a> from the the Set returned by getParsedSamples() for a given id.</td>
</tr>
<tr>
<td align="left">getParsedSamples()</td>
<td align="left">Return 1 <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/ParsedSample.html">ParsedSample</a> for each classified sample in the dataset.</td>
</tr>
</tbody>
</table>
<h4 id="otunodeimpl-is-the-superclass-for-the-biolockjnode-package"><a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/OtuNodeImpl.html">OtuNodeImpl</a> is the superclass for the <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/package-summary.html">biolockj.node</a> package.</h4>
<table>
<thead>
<tr>
<th align="left">Method</th>
<th align="left">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">addOtu(level, otu)</td>
<td align="left">A node represents a single OTU, each level in the taxonomic hierarchy is populated with this method.</td>
</tr>
<tr>
<td align="left">getCount()</td>
<td align="left">Get the OTU count.</td>
</tr>
<tr>
<td align="left">getLine()</td>
<td align="left">Get the classifier report line used to create the node.</td>
</tr>
<tr>
<td align="left">getOtuMap()</td>
<td align="left">This map may contain 1 element for each of the <em>report.taxonomyLevels</em> and is populated by addOtu(level, otu).</td>
</tr>
<tr>
<td align="left">getSampleId()</td>
<td align="left">Get the sample ID to which the OTU belongs.</td>
</tr>
<tr>
<td align="left">report()</td>
<td align="left">Print node info to log file as DEBUG line - not visible unless <em>pipeline.logLevel=DEBUG</em>.</td>
</tr>
<tr>
<td align="left">setCount(num)</td>
<td align="left">Set the OTU count.</td>
</tr>
<tr>
<td align="left">setLine(line)</td>
<td align="left">Set the classifier report line used to create the node.</td>
</tr>
<tr>
<td align="left">setSampleId(id)</td>
<td align="left">set the sample ID to which the OTU belongs.</td>
</tr>
</tbody>
</table>
<ul>
<li><a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/OtuNodeImpl.html">OtuNodeImpl</a> methods do not need to be overridden.  </li>
<li>New <a href="https://msioda.github.io/BioLockJ/docs/biolockj/node/OtuNode.html">OtuNode</a> implementations should call existing methods from their constructor.</li>
</ul>
<h2 id="share-your-module">Share your module</h2>
<p>Other users can use your module by putting the compiled class in the java classpath when they run BioLockj and using #BioModule <strong>classname</strong> in their config file.</p>
<p>You can share your module as code or as a compiled class any way that you like.  The official repository for external BioLockJ modules is <a href="https://github.com/IvoryC/blj_ext_modules">blj_ext_modules</a>.  Each module has a folder at the top level of the repository and should include the java code as well a config file to test the module alone, a test file to run a multi-module pipeline that includes the module, and (where applicable) a dockerfile.</p>
  
    
  

              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../Validation/" title="Validation" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Validation
              </span>
            </div>
          </a>
        
        
          <a href="../Example-Pipeline/" title="Example Pipeline" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Example Pipeline
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            2019 Fodor Lab
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.8e03edeb.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
    
  </body>
</html>